File: Exemplos/App_eXperience_ReactNative/Elgin_AppExperience_ReactNative-font/E1_M8_ReactNative/E1_M8_ReactNative_Font/android/app/src/main/java/com/elginm8/ServiceSat.java
package com.elginm8;

import android.content.Context;
import android.os.Build;
import android.os.Environment;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.RequiresApi;

import java.io.File;
import java.io.FileWriter;

import java.util.Arrays;
import java.util.Optional;
import java.util.Random;

import java.util.concurrent.atomic.AtomicReference;

import br.com.elgin.DeviceInfo;
import br.com.elgin.Sat;
import br.com.elgin.SatInitializer;
import com.facebook.react.bridge.ReadableMap;

public class ServiceSat {
    public static SatInitializer satInitializer;
    private static Context contextSat;

    private static String BASE_ROOT_DIR;

    private static final String SATLOG_ARCHIVE_NAME = "SatLog.txt";

    public  ServiceSat (Context context){
        satInitializer = new SatInitializer();
        contextSat = context;

        BASE_ROOT_DIR = "/Android/data/" + contextSat.getPackageName() + "/";
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static String ativarSAT(ReadableMap map){
        int numSessao = getNumeroSessao();
        int subComando = map.getInt("subComando");
        String codeAtivacao = map.getString("codeAtivacao");
        String cnpj = map.getString("cnpj");
        int cUF = map.getInt("cUF");

        StringBuilder valueRetorn = new StringBuilder(Sat.ativarSat(numSessao, subComando, codeAtivacao, cnpj, cUF));
        if (valueRetorn.toString().contains("NomParsing")) {
            valueRetorn = new StringBuilder(valueRetorn.toString().replace("NomParsing(", "").replace(", Digit)", ""));

            int[] arr = Arrays.stream(valueRetorn.substring(1, valueRetorn.length() - 1).split(","))
                    .map(String::trim).mapToInt(Integer::parseInt).toArray();

            valueRetorn = new StringBuilder();
            for (int item : arr) {
                valueRetorn.append((char) item);
            }
        }

        mostrarRetorno(valueRetorn.toString());
        return valueRetorn.toString();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static String associarAssinatura(ReadableMap map) {
        int numSessao = getNumeroSessao();
        String codeAtivacao = map.getString("codeAtivacao");
        String cnpjSh = map.getString("cnpjSh");
        String assinaturaAC = map.getString("assinaturaAC");

        StringBuilder valueRetorn = new StringBuilder(Sat.associarAssinatura(numSessao, codeAtivacao, cnpjSh, assinaturaAC));
        if (valueRetorn.toString().contains("NomParsing")) {
            valueRetorn = new StringBuilder(valueRetorn.toString().replace("NomParsing(", "").replace(", Digit)", ""));

            int[] arr = Arrays.stream(valueRetorn.substring(1, valueRetorn.length() - 1).split(","))
                    .map(String::trim).mapToInt(Integer::parseInt).toArray();

            valueRetorn = new StringBuilder();
            for (int item : arr) {
                valueRetorn.append((char) item);
            }
        }

        mostrarRetorno(valueRetorn.toString());
        return valueRetorn.toString();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static String consultarSAT(ReadableMap map) {
        int numSessao = getNumeroSessao();

        StringBuilder valueRetorn = new StringBuilder(Sat.consultarSat((Integer) numSessao));
        if (valueRetorn.toString().contains("NomParsing")) {
            valueRetorn = new StringBuilder(valueRetorn.toString().replace("NomParsing(", "").replace(", Digit)", ""));

            int[] arr = Arrays.stream(valueRetorn.substring(1, valueRetorn.length() - 1).split(","))
                    .map(String::trim).mapToInt(Integer::parseInt).toArray();

            valueRetorn = new StringBuilder();
            for (int item : arr) {
                valueRetorn.append((char) item);
            }
        }

        mostrarRetorno(valueRetorn.toString());
        return valueRetorn.toString();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static String statusOperacional(ReadableMap map){
        int numSessao = getNumeroSessao();
        String codeAtivacao = map.getString("codeAtivacao");

        StringBuilder valueRetorn = new StringBuilder(Sat.consultarStatusOperacional(numSessao,codeAtivacao));
        if (valueRetorn.toString().contains("NomParsing")) {
            valueRetorn = new StringBuilder(valueRetorn.toString().replace("NomParsing(", "").replace(", Digit)", ""));

            int[] arr = Arrays.stream(valueRetorn.substring(1, valueRetorn.length() - 1).split(","))
                    .map(String::trim).mapToInt(Integer::parseInt).toArray();

            valueRetorn = new StringBuilder();
            for (int item : arr) {
                valueRetorn.append((char) item);
            }
        }

        mostrarRetorno(valueRetorn.toString());

        return valueRetorn.toString();
    }


    @RequiresApi(api = Build.VERSION_CODES.N)
    public static String enviarVenda(ReadableMap map){
        int numSessao = getNumeroSessao();
        String codeAtivacao = map.getString("codeAtivacao");
        String xmlSale = map.getString("xmlSale");

        StringBuilder valueRetorn = new StringBuilder(Sat.enviarDadosVenda(numSessao, codeAtivacao, xmlSale));
        if (valueRetorn.toString().contains("NomParsing")) {
            valueRetorn = new StringBuilder(valueRetorn.toString().replace("NomParsing(", "").replace(", Digit)", ""));

            int[] arr = Arrays.stream(valueRetorn.substring(1, valueRetorn.length() - 1).split(","))
                    .map(String::trim).mapToInt(Integer::parseInt).toArray();

            valueRetorn = new StringBuilder();
            for (int item : arr) {
                valueRetorn.append((char) item);
            }
        }

        mostrarRetorno(valueRetorn.toString());
        return valueRetorn.toString();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static String cancelarVenda(ReadableMap map){
        int numSessao = getNumeroSessao();
        String codeAtivacao = map.getString("codeAtivacao");
        String cFeNumber = map.getString("cFeNumber");
        String xmlCancelamento = map.getString("xmlCancelamento");

        StringBuilder valueRetorn = new StringBuilder(Sat.cancelarUltimaVenda(numSessao, codeAtivacao, cFeNumber, xmlCancelamento));
        if (valueRetorn.toString().contains("NomParsing")) {
            valueRetorn = new StringBuilder(valueRetorn.toString().replace("NomParsing(", "").replace(", Digit)", ""));

            int[] arr = Arrays.stream(valueRetorn.substring(1, valueRetorn.length() - 1).split(","))
                    .map(String::trim).mapToInt(Integer::parseInt).toArray();

            valueRetorn = new StringBuilder();
            for (int item : arr) {
                valueRetorn.append((char) item);
            }
        }

        mostrarRetorno(valueRetorn.toString());
        return valueRetorn.toString();
    }

    public static String extrairLog(ReadableMap map){
        int numSessao = (Integer) map.getInt("numSessao");
        String codeAtivacao = (String) map.getString("codeAtivacao");

        String extractedLog = "";
        extractedLog = Sat.extrairLogs(numSessao,codeAtivacao);

        if(extractedLog.equals("DeviceNotFound")){
            return extractedLog;
        }

        String extractedLogInArray[] = extractedLog.split("[|]");

        byte[] byteArrayWithStringConverted = android.util.Base64.decode(extractedLogInArray[5],android.util.Base64.DEFAULT);

        String logtoBeSavedInStorage = new String(byteArrayWithStringConverted);
        System.out.println("LOG SAT:"+logtoBeSavedInStorage);

        return logtoBeSavedInStorage;
    }


    private static void mostrarRetorno(String retorno) {
        Toast.makeText(contextSat, String.format("Retorno: %s", retorno), Toast.LENGTH_LONG).show();
    }

    private static int getNumeroSessao() {
        return new Random().nextInt(1_000_000);
    }

    private static String storagePathToAbsolutePath(String pathFromStorageRoot) {
        final StringBuffer buffer = new StringBuffer(255);

        buffer.append(Environment.getExternalStorageDirectory().getAbsolutePath());

        if (pathFromStorageRoot.charAt(0) != '/') {
            buffer.append('/');
        }

        buffer.append(pathFromStorageRoot);

        return buffer.toString();
    }

    private static boolean createRootDirectory(Context ctx) {
        // Constant, copied here: private static final String BASE_DIR = "/Android/data/";
        final String dataDir = storagePathToAbsolutePath(BASE_ROOT_DIR);
        System.out.println("SAT DIR"+dataDir);
        System.out.println("SAT DIR"+storagePathToAbsolutePath(BASE_ROOT_DIR));

        final File f = new File(dataDir);
        return f.mkdirs();
    }
    private static void writeFileOnStorage(String fileNameWithExtension, String textToWrite){
        File file = new File(storagePathToAbsolutePath(BASE_ROOT_DIR), "files");


        if(!file.exists()){
            file.mkdir();
        }

        try{
            File gpxfile = new File(file, fileNameWithExtension);
            FileWriter writer = new FileWriter(gpxfile);
            writer.append(textToWrite);
            writer.flush();
            writer.close();

        }catch (Exception e){
            e.printStackTrace();
        }
    }
}

File: Exemplos/App_eXperience_XamarinAndroid/Elgin_AppExperience_XamarinAndroid-font/Sat/ServiceSat.cs
using Android.Content;
using Android.Widget;
using System.Collections.Generic;
using BR.Com.Elgin;
using Android.Util;
using Java.Lang;
using Android.OS;
using Java.IO;

namespace M8
{
    class ServiceSat
    {
        private readonly Context contextSat;
        public static SatInitializer satInitializer;

        //Diretório Raiz da aplicação
        private string BASE_ROOT_DIR;

        private const string SATLOG_ARCHIVE_NAME = "SatLog.txt";

        public ServiceSat(Context context)
        {
            satInitializer = new SatInitializer();
            contextSat = context;

            BASE_ROOT_DIR = "/Android/data/" + contextSat.PackageName + "/";
        }

        public string AtivarSAT(Dictionary<string, object> map)
        {
            int numSessao = (int)map["numSessao"];
            int subComando = (int)map["subComando"];
            string codeAtivacao = (string)map["codeAtivacao"];
            string cnpj = (string)map["cnpj"];
            int cUF = (int) map["cUF"];
            string retorno = Sat.AtivarSat(numSessao, subComando, codeAtivacao, cnpj, cUF);
            MostrarRetorno(retorno);
            return retorno;
        }

        public string AssociarAssinatura(Dictionary<string, object> map)
        {
            int numSessao = (int)map["numSessao"];
            string codeAtivacao = (string)map["codeAtivacao"];
            string cnpjSh = (string) map["cnpjSh"];
            string assinaturaAC = (string)map["assinaturaAC"];

            string retorno = Sat.AssociarAssinatura(numSessao, codeAtivacao, cnpjSh, assinaturaAC);
            MostrarRetorno(retorno);
            return retorno;
        }

        public string ConsultarSAT(Dictionary<string, object> map)
        {
            int numSessao = (int) map["numSessao"];

            string result = Sat.ConsultarSat(numSessao);
            MostrarRetorno(result);
            return result;
        }

        public string StatusOperacional(Dictionary<string, object> map)
        {
            int numSessao = (int)map["numSessao"];
            string codeAtivacao = (string)map["codeAtivacao"];

            string retorno = Sat.ConsultarStatusOperacional(numSessao, codeAtivacao);
            MostrarRetorno(retorno);
            return retorno;
        }

        public string EnviarVenda(Dictionary<string, object> map)
        {
            int numSessao = (int) map["numSessao"];
            string codeAtivacao = (string) map["codeAtivacao"];
            string xmlSale = (string) map["xmlSale"];

            string retorno = Sat.EnviarDadosVenda(numSessao, codeAtivacao, xmlSale);
            MostrarRetorno(retorno);
            return retorno;
        }

        public string CancelarVenda(Dictionary<string, object> map)
        {
            int numSessao = (int) map["numSessao"];
            string codeAtivacao = (string)map["codeAtivacao"];
            string cFeNumber = (string)map["cFeNumber"];
            string xmlCancelamento = (string) map["xmlCancelamento"];

            string retorno = Sat.CancelarUltimaVenda(numSessao, codeAtivacao, cFeNumber, xmlCancelamento);
            MostrarRetorno(retorno);
            return retorno;
        }

        private void MostrarRetorno(string retorno)
        {
            Toast.MakeText(contextSat, string.Format("Retorno: {0}", retorno), ToastLength.Long).Show();
        }

        public bool ExtrairLog(Dictionary<string,object> map)
        {
            int numSessao = (int)map["numSessao"];
            string codeAtivacao = (string)map["codeAtivacao"];

            string extractedLog = Sat.ExtrairLogs(numSessao, codeAtivacao);

            //Se o sat não estiver conectado ou algum outro problema tiver ocorrido, o retorno de Sat.extrairLogs()  será 'DeviceNotFound' e função não foi bem sucedida!
            if (extractedLog.Equals("DeviceNotFound"))
            {
                MostrarRetorno(extractedLog);
                return false;
            }

            //O texto a ser salvo está na 6° posição ao separar a String por '|'
            string[] extractedLogInArray = extractedLog.Split("|");

            //O texto a ser salvo está em base64 e deve ser decodificado antes de ser salvo no arquivo .txt
            byte[] byteArrayWithStringConverted = Base64.Decode(extractedLogInArray[5], Base64Flags.Default);

            String logtoBeSavedInStorage = new String(byteArrayWithStringConverted);

            //Tenta criar o dir ROOT onde a aplicação irá conseguir salvar o arquivo do logsat
            CreateRootDirectory(contextSat);
            WriteFileOnStorage(SATLOG_ARCHIVE_NAME, logtoBeSavedInStorage);

            return true;
        }

        //Função que retorna o PATH absoluto do storage do Android, utilizada para criar o diretório root caso não exista
        //e para prover o PATH correto onde escrever o logSAT extraído
        private static string StoragePathToAbsolutePath(string pathFromStorageRoot)
        {
            StringBuffer buffer = new StringBuffer(255);

            buffer.Append(Environment.ExternalStorageDirectory.AbsolutePath);

            if (pathFromStorageRoot[0] != '/')
            {
                buffer.Append('/');
            }

            buffer.Append(pathFromStorageRoot);

            return buffer.ToString();
        }

        //Função que cria o diretório root da aplicação (com.packagename) localizado em Android/data
        //diretório que será utilizado para salvar o arquivo de log do sat
        private bool CreateRootDirectory(Context ctx)
        {
            // Constant, copied here: private static final String BASE_DIR = "/Android/data/";
            string dataDir = StoragePathToAbsolutePath(BASE_ROOT_DIR);
            Log.Debug("MADARA", dataDir);
            Log.Debug("MADARA", StoragePathToAbsolutePath(BASE_ROOT_DIR));
            File f = new File(dataDir);
            return f.Mkdirs();
        }

        private void WriteFileOnStorage(string fileNameWithExtension, String textToWrite)
        {
            File file = new File(StoragePathToAbsolutePath(BASE_ROOT_DIR), "files");
            if (!file.Exists())
            {
                file.Mkdir();
            }

            try
            {
                File gpxfile = new File(file, fileNameWithExtension);
                FileWriter writer = new FileWriter(gpxfile);
                writer.Append(textToWrite);
                writer.Flush();
                writer.Close();

                Toast.MakeText(contextSat, "Saved your text in " + gpxfile.Path, ToastLength.Long).Show();
            }
            catch (Exception e)
            {
                e.PrintStackTrace();
            }
        }

        public string GetBASE_ROOT_DIR()
        {
            return BASE_ROOT_DIR;
        }
    }
}